From f8aeb55bea64fde80a63a88d5d7bad8fd2dde3d9 Mon Sep 17 00:00:00 2001
From: Alex Deryskyba <alex@codesnake.com>
Date: Sun, 2 Nov 2014 15:54:29 +0200
Subject: [PATCH 15/17] [aml] Add some key mappings for HDMI CEC

---
 xbmc/input/linux/LinuxInputDevices.cpp |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/xbmc/input/linux/LinuxInputDevices.cpp b/xbmc/input/linux/LinuxInputDevices.cpp
index 0b0f7ae..c84930f 100644
--- a/xbmc/input/linux/LinuxInputDevices.cpp
+++ b/xbmc/input/linux/LinuxInputDevices.cpp
@@ -260,10 +260,10 @@ KeyMap keyMap[] = {
   { KEY_ZOOM          , XBMCK_ZOOM        },
   { KEY_TEXT          , XBMCK_TEXT        },
   { KEY_FAVORITES     , XBMCK_FAVORITES   },
-  { KEY_RED           , XBMCK_RED         },
-  { KEY_GREEN         , XBMCK_GREEN       },
-  { KEY_YELLOW        , XBMCK_YELLOW      },
-  { KEY_BLUE          , XBMCK_BLUE        },
+  { KEY_RED           , XBMCK_TAB         },
+  { KEY_GREEN         , XBMCK_z           },
+  { KEY_YELLOW        , XBMCK_i           },
+  { KEY_BLUE          , XBMCK_c           },
   { KEY_HOMEPAGE      , XBMCK_HOMEPAGE    },
   { KEY_MAIL          , XBMCK_LAUNCH_MAIL },
   { KEY_SEARCH        , XBMCK_BROWSER_SEARCH},
-- 
1.7.10.4

